src/hooks/useSignChallenge.ts

```diff
@@ -0,0 +1,116 @@
+import { useState, useCallback } from 'react';
+import { useSignMessage } from 'wagmi';
+import { SIGNING_SERVICE_URL } from '../constants/constants';
+import { storageKeys } from '../constants/localStorage';
+import { SiweMessage } from 'siwe';
+
+export interface SiweSignatureData {
+  signatureSet: boolean;
+  expirationDate: string;
+}
+
+export function useSiweSignature(address?: `0x${string}`) {
+  const { signMessageAsync } = useSignMessage();
+  const [requiresSign, setRequiresSign] = useState(false);
+
+  // Used to wait for the modal interaction and/or return of the
+  // signing promise.
+  const [signPromise, setSignPromise] = useState<{
+    resolve: () => void;
+    reject: (reason: Error) => void;
+  } | null>(null);
+
+  const storageKey = `${storageKeys.SIWE_SIGNATURE_KEY_PREFIX}${address}`;
+
+  const checkStoredSignature = useCallback((): SiweSignatureData | null => {
+    if (!address) return null;
+
+    try {
+      const stored = localStorage.getItem(storageKey);
+      if (!stored) return null;
+
+      const data: SiweSignatureData = JSON.parse(stored);
+      return new Date(data.expirationDate) > new Date() ? data : null;
+    } catch {
+      localStorage.removeItem(storageKey);
+      return null;
+    }
+  }, [address, storageKey]);
+
+  const signMessage = useCallback((): Promise<void> => {
+    return new Promise((resolve, reject) => {
+      setSignPromise({ resolve, reject });
+      setRequiresSign(true);
+    });
```

Comment:
I think storing this promise is not necessary and I would avoid it. If you want to have the `handleSign` adjust based on a pending signature dialog or not, then it should be enough to check for `if (!requiresSign)` in l. 48, no?

Reply:
I tried refactoring the code and while I could reduce parts of the `useSiweSignature` code overall, it didn't necessarily turn out any less complex or easier to read. So I guess we just have to live with some confusion/complexity around this part. I'm fine with keeping it the way you implemented it 👍

Reply:
Okay! Out of curious, what did you attempted to remove the promise? `useEffect` ? Yeah the whole complexity fact is that the process is done inside the `sep10` function, sometimes is hard to mix hook and non hook processes like this.